stl/inc/format

```diff
@@ -398,26 +570,256 @@ public:
+
+public:
+    constexpr _Numeric_specs_checker(_Type _Arg_type) : _Arg_type(_Arg_type) {}
+
+    constexpr void _Require_numeric_argument() {
+        if (!_Is_arithmetic_fmt_type(_Arg_type)) {
+            throw format_error("Format specifier requires numeric argument.");
+        }
+    }
+
+    constexpr void _Check_sign() {
+        _Require_numeric_argument();
+        if (_Is_integral_fmt_type(_Arg_type) && _Arg_type != _Type::_Int32 && _Arg_type != _Type::_Int64
+            && _Arg_type != _Type::_Char) {
+            throw format_error("Format specifier requires signed argument.");
+        }
+    }
+
+    constexpr void _Check_precision() {
+        if (_Is_integral_fmt_type(_Arg_type) || _Arg_type == _Type::_Pointer) {
+            throw format_error("Precision not allowed for this argument type.");
+        }
+    }
+};
+
+// Uses _Numeric_specs_checker to check that the type of the argument printed by
+// a replacement field with format specs actually satisfies the requirements for
+// that format spec. If the requirements are met then calls the base class
+// handler method.
+template <class _Handler>
+class _Specs_checker : public _Handler {
+    _Numeric_specs_checker _Numeric_checker;
+
+public:
+    constexpr _Specs_checker(const _Handler& _Handler_inst, _Type _Arg_type)
+        : _Handler(_Handler_inst), _Numeric_checker(_Arg_type) {}
+
+    // _On_align has no checking, since we don't implement numeric alignments.
+    constexpr void _On_sign(_Sign _Sgn) {
+        _Numeric_checker._Check_sign();
+        _Handler::_On_sign(_Sgn);
+    }
+
+    constexpr void _On_hash() {
+        _Numeric_checker._Require_numeric_argument();
+        _Handler::_On_hash();
+    }
+
+    constexpr void _On_zero() {
+        _Numeric_checker._Require_numeric_argument();
+        _Handler::_On_zero();
+    }
+
+    constexpr void _On_precision(unsigned int _Precision) {
+        _Numeric_checker._Check_precision();
+        _Handler::_On_precision(_Precision);
+    }
+};
+
+template <class _Ty, class _CharT = char>
+struct formatter;
+
+
 // TODO: test coverage
 template <class _Context, class... _Args>
 class _Format_arg_store {
     static constexpr size_t _Num_args = sizeof...(_Args);
     // TODO: no arg packing yet
-
     using _Value_type = basic_format_arg<_Context>;
+    _STD array<basic_format_arg<_Context>, _Num_args> _Arg_array;
 };
 
 // TODO: test coverage
 template <class _Context>
 class basic_format_args {
+    size_t _Size;
+    const basic_format_arg<_Context>* _Data;
+
 public:
     basic_format_args() noexcept;
     template <class... _Args>
     basic_format_args(const _Format_arg_store<_Context, _Args...>& store) noexcept;
 
-    basic_format_arg<_Context> get(size_t _Index) const noexcept;
+    basic_format_arg<_Context> get(size_t _Index) const noexcept {
+        if (_Index < _Size) {
+            return _Data[_Index];
+        } else {
+            return basic_format_arg<_Context>();

```

Comment:
In new code we try to use empty braces to construct temporaries (as this avoids looking like a function call).